### DISCUSSION TITLE
v1.20.0

### DISCUSSION THREAD

**Discussion by kyleconroy** [sqlc-dev/sqlc#2550] [Announcements]
### `kyleconroy/sqlc` is now `sqlc-dev/sqlc`

We've completed our migration to the [sqlc-dev/sqlc](https://github.com/sqlc-dev/sqlc) repository. All existing links and installation instructions will continue to work. If you're using the `go` tool to install `sqlc`, you'll need to use the new import path to get v1.20.0 (and all future versions).

```sh
# INCORRECT: old import path
go install github.com/kyleconroy/sqlc/cmd/sqlc@v1.20.0

# CORRECT: new import path
go install github.com/sqlc-dev/sqlc/cmd/sqlc@v1.20.0
```

We designed the upgrade process to be as smooth as possible. If you run into any issues, please [file a bug report](https://github.com/sqlc-dev/sqlc/issues/new?assignees=&labels=bug%2Ctriage&projects=&template=BUG_REPORT.yml) via GitHub.

### Use `EXPLAIN ...` output in lint rules

`sqlc vet` can now run `EXPLAIN` on your queries and include the results for use in your lint rules. For example, this rule checks that `SELECT` queries use an index.

```yaml
version: 2
sql:
  - schema: "query.sql"
    queries: "query.sql"
    engine: "postgresql"
    database:
      uri: "postgresql://postgres:postgres@localhost:5432/postgres"
    gen:
      go:
        package: "db"
        out: "db"
    rules:
      - has-index
rules:
- name: has-index
  rule: >
    query.sql.startsWith("SELECT") &&
    !(postgresql.explain.plan.plans.all(p, has(p.index_name) || p.plans.all(p, has(p.index_name))))
```

The expression environment has two variables containing `EXPLAIN ...` output, `postgresql.explain` and `mysql.explain`. `sqlc` only populates the variable associated with your configured database engine, and only when you have a [database connection configured](https://docs.sqlc.dev/en/stable/reference/config.html#database).

For the `postgresql` engine, `sqlc` runs

```sql
EXPLAIN (ANALYZE false, VERBOSE, COSTS, SETTINGS, BUFFERS, FORMAT JSON) ...
```

where `"..."` is your query string, and parses the output into a [`PostgreSQLExplain`](https://buf.build/sqlc/sqlc/docs/v1.20.0:vet#vet.PostgreSQLExplain) proto message.

For the `mysql` engine, `sqlc` runs

```sql
EXPLAIN FORMAT=JSON ...
```

where `"..."` is your query string, and parses the output into a [`MySQLExplain`](https://buf.build/sqlc/sqlc/docs/v1.20.0:vet#vet.MySQLExplain) proto message.

These proto message definitions are too long to include here, but you can find them in the `protos` directory within the `sqlc` source tree.

The output from `EXPLAIN ...` depends on the structure of your query so it's a bit difficult to offer generic examples. Refer to the [PostgreSQL documentation](https://www.postgresql.org/docs/current/using-explain.html) and [MySQL documentation](https://dev.mysql.com/doc/refman/en/explain-output.html) for more information.

```yaml
...
rules:
- name: postgresql-query-too-costly
  message: "Query cost estimate is too high"
  rule: "postgresql.explain.plan.total_cost > 1.0"
- name: postgresql-no-seq-scan
  message: "Query plan results in a sequential scan"
  rule: "postgresql.explain.plan.node_type == 'Seq Scan'"
- name: mysql-query-too-costly
  message: "Query cost estimate is too high"
  rule: "has(mysql.explain.query_block.cost_info) && double(mysql.explain.query_block.cost_info.query_cost) > 2.0"
- name: mysql-must-use-primary-key
  message: "Query plan doesn't use primary key"
  rule: "has(mysql.explain.query_block.table.key) && mysql.explain.query_block.table.key != 'PRIMARY'"
```

When building rules that depend on `EXPLAIN ...` output, it may be helpful to see the actual JSON returned from the database. `sqlc` will print it When you set the environment variable `SQLCDEBUG=dumpexplain=1`. Use this environment variable together with a dummy rule to see `EXPLAIN ...` output for all of your queries.

### Opting-out of lint rules

For any query, you can tell `sqlc vet` not to evaluate lint rules using the `@sqlc-vet-disable` query annotation.

```sql
/* name: GetAuthor :one */
/* @sqlc-vet-disable */
SELECT * FROM authors
WHERE id = ? LIMIT 1;
```

### Bulk insert for MySQL

_Developed by [@Jille](https://github.com/Jille)_

MySQL now supports the `:copyfrom` query annotation. The generated code uses the [LOAD DATA](https://dev.mysql.com/doc/refman/8.0/en/load-data.html) command to insert data quickly and efficiently.

Use caution with this feature. Errors and duplicate keys are treated as warnings and insertion will continue, even without an error for some cases.  Use this in a transaction and use `SHOW WARNINGS` to check for any problems and roll back if necessary.

Check the [error handling](https://dev.mysql.com/doc/refman/8.0/en/load-data.html#load-data-error-handling) documentation for more information.

```sql
CREATE TABLE foo (a text, b integer, c DATETIME, d DATE);

-- name: InsertValues :copyfrom
INSERT INTO foo (a, b, c, d) VALUES (?, ?, ?, ?);
```

```go
func (q *Queries) InsertValues(ctx context.Context, arg []InsertValuesParams) (int64, error) {
	...
}
```

`LOAD DATA` support must be enabled in the MySQL server.

### CAST support for MySQL

_Developed by [@ryanpbrewster](https://github.com/ryanpbrewster) and [@RadhiFadlillah](https://github.com/RadhiFadlillah)_

`sqlc` now understands `CAST` calls in MySQL queries, offering greater flexibility when generating code for complex queries.

```sql
CREATE TABLE foo (bar BOOLEAN NOT NULL);

-- name: SelectColumnCast :many
SELECT CAST(bar AS BIGINT) FROM foo;
```

```go
package querytest

import (
	"context"
)

const selectColumnCast = `-- name: SelectColumnCast :many
SELECT CAST(bar AS BIGINT) FROM foo
`

func (q *Queries) SelectColumnCast(ctx context.Context) ([]int64, error) {
  ...
}
```

### SQLite improvements

AA slew of fixes landed for our SQLite implementation, bringing it closer to parity with MySQL and PostgreSQL. We want to thank [@orisano](https://github.com/orisano) for their continued dedication to improving `sqlc`'s SQLite support.

## What's Changed
* Mention UUID type for pgx/v5 by @peterwillis in https://github.com/sqlc-dev/sqlc/pull/2448
* Update BUG_REPORT.yml by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2460
* scripts: Add new version to BUG_REPORT.yml by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2461
* build: Run all supported versions of MySQL / PostgreSQL by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2463
* docs: Add a description to the document for cases when a query result has no rows by @n3xem in https://github.com/sqlc-dev/sqlc/pull/2462
* feat: add basic parsing of boolean flag query metadata annotations by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2464
* feat(vet): Introduce a query annotation to opt out of sqlc vet rules by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2474
* fix(vet): report an error when a query is unpreparable, remove dead code by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2486
* build(deps): bump pygments from 2.7.4 to 2.15.0 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2485
* build(deps): bump github.com/jackc/pgconn from 1.14.0 to 1.14.1 by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2483
* feat(vet): Add output from `EXPLAIN ...` for queries to the CEL program environment by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2489
* docs: update copyright and author by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2490
* fix: sqlite log by @mcdoker18 in https://github.com/sqlc-dev/sqlc/pull/2488
* build(deps): bump github.com/google/cel-go from 0.16.0 to 0.17.1 by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2484
* docs: Check Python dependencies via dependabot by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2497
* build(deps): bump idna from 2.10 to 3.4 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2499
* build(deps): bump packaging from 20.9 to 23.1 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2498
* build(deps): bump pygments from 2.15.0 to 2.15.1 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2500
* fix(engine/sqlite): support for repeated table_option by @orisano in https://github.com/sqlc-dev/sqlc/pull/2482
* test: Add profiles to docker compose by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2503
* Add example sqlc.yaml for migration parsing by @samfweb in https://github.com/sqlc-dev/sqlc/pull/2479
* fix(engine/sqlite): added function to convert cast node by @orisano in https://github.com/sqlc-dev/sqlc/pull/2470
* build(deps): bump certifi from 2022.12.7 to 2023.7.22 in /docs by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2504
* fix(engine/sqlite): fixed to be able to find relation from WITH clause by @orisano in https://github.com/sqlc-dev/sqlc/pull/2444
* build(deps): bump sphinx from 4.4.0 to 6.1.0 in /docs by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2505
* build: Add psql and mysqlsh to devenv by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2507
* fix(sql/catalog): Support pg_dump output by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2508
* docs: small updates by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2506
* fix(engine/sqlite): fixed IN operator precedence by @orisano in https://github.com/sqlc-dev/sqlc/pull/2428
* fix(engine/sqlite): fix table_alias rules by @orisano in https://github.com/sqlc-dev/sqlc/pull/2465
* fix(engine/sqlite): fix join_operator rule by @orisano in https://github.com/sqlc-dev/sqlc/pull/2434
* fix(vet): clean up unnecessary `prepareable()` func and a var name collision by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2509
* Clarify option package support by @johanbrandhorst in https://github.com/sqlc-dev/sqlc/pull/2513
* fix(vet): Split vet messages out of codegen.proto by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2511
* build(deps): bump urllib3 from 1.26.5 to 2.0.4 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2516
* build(deps): bump chardet from 4.0.0 to 5.1.0 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2517
* build(deps): bump snowballstemmer from 2.1.0 to 2.2.0 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2519
* build(deps): bump pytz from 2021.1 to 2023.3 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2520
* build(deps): bump sphinxcontrib-htmlhelp from 2.0.0 to 2.0.1 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2518
* fix missing unsigned param by @hiroebe in https://github.com/sqlc-dev/sqlc/pull/2522
* feat(debug): add debug flag and docs for dumping vet rule variables by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2521
* Fix code geneation for sqlc.slice by @SoMuchForSubtlety in https://github.com/sqlc-dev/sqlc/pull/2431
* build(deps): bump pyparsing from 2.4.7 to 3.1.0 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2530
* build(deps): bump alabaster from 0.7.12 to 0.7.13 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2526
* feat(sqlite): virtual tables and fts5 supported by @mcdoker18 in https://github.com/sqlc-dev/sqlc/pull/2531
* fix(vet): query.cmd was always set to ":" by @thokra-nav in https://github.com/sqlc-dev/sqlc/pull/2525
* build(docs): Ignore updates for sphinx by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2532
* build(deps): bump babel from 2.9.1 to 2.12.1 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2527
* build(deps): bump sphinxcontrib-applehelp from 1.0.2 to 1.0.4 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2533
* chore!: Rename kyleconroy/sqlc to sqlc-dev/sqlc by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2523
* docs: Point GitHub links to new repository location by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2534
* fix(engine/sqlite): To lowercase ast.ResTarget.Name by @orisano in https://github.com/sqlc-dev/sqlc/pull/2433
* chore(proto): reformat protos using `buf format -w` by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2536
* build(deps): bump google.golang.org/grpc from 1.56.2 to 1.57.0 by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2535
* fix(compiler): correctly validate alias in order/group by clauses for joins by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2537
* chore: Update FEATURE_REQUEST.yml to include SQLite engine option by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2540
* feat(sql/catalog): Support ALTER TABLE IF EXISTS by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2542
* feat(mysql): Implement cast function parser by @RadhiFadlillah in https://github.com/sqlc-dev/sqlc/pull/2473
* feat(mysql): :copyfrom support via LOAD DATA INFILE by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2545
* feat(postgresql): Add support for PostgreSQL multi-dimensional arrays by @colli173 in https://github.com/sqlc-dev/sqlc/pull/2338
* fix(codegen/golang): fix sqlc.embed to work with pq.Array wrap by @orisano in https://github.com/sqlc-dev/sqlc/pull/2544
* refactor(compiler): remove some duplicate code by @andrewmbenton in https://github.com/sqlc-dev/sqlc/pull/2546
* build(deps): bump pyparsing from 3.1.0 to 3.1.1 in /docs by @dependabot in https://github.com/sqlc-dev/sqlc/pull/2547
* chore: Finish migration to sqlc-dev/sqlc by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2548
* feat(vet): Add default query parameters for explain queries by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2543
* feat(cmd/sqlc): Bump version to 1.20.0 by @kyleconroy in https://github.com/sqlc-dev/sqlc/pull/2549

## New Contributors
* @peterwillis made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2448
* @n3xem made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2462
* @samfweb made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2479
* @hiroebe made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2522
* @SoMuchForSubtlety made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2431
* @thokra-nav made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2525
* @RadhiFadlillah made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2473
* @colli173 made their first contribution in https://github.com/sqlc-dev/sqlc/pull/2338

**Full Changelog**: https://github.com/sqlc-dev/sqlc/compare/v1.19.1...v1.20.0

<hr /><em>This discussion was created from the release <a href='https://github.com/sqlc-dev/sqlc/releases/tag/v1.20.0'>v1.20.0</a>.</em>

**Comment by fresonn** [comment]
Could you publish a new version on snap
